type-generator/src/main/java/io/mesosphere/types/dtr/compiler/impl/StringTemplateEngine.java

```diff
@@ -0,0 +1,777 @@
+import io.mesosphere.types.dtr.models.internal.types.*;
+import io.mesosphere.types.dtr.models.internal.values.AnyValue;
+import io.mesosphere.types.dtr.models.internal.values.DateTimeValue;
+import io.mesosphere.types.dtr.repository.internal.RepositoryView;
+import org.antlr.runtime.Token;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.StringEscapeUtils;
+import org.ini4j.Ini;
+import org.stringtemplate.v4.ST;
+import org.stringtemplate.v4.STGroup;
+import org.stringtemplate.v4.STGroupString;
+import org.stringtemplate.v4.StringRenderer;
+import org.stringtemplate.v4.misc.Misc;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.TemporalAccessor;
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * A StringTemplateEngine engine is using the StringTemplateEngine template engine to
+ * compile a user expression into a buffer.
+ */
+public class StringTemplateEngine implements CompilerEngine {
+
+    /**
+     * The zone ID to use when parsing/writing dates
+     */
+    private ZoneId dateZone;
+
+    /**
+     * Mapping of scalar types
+     */
+    private Map<String, String> scalarTypes;
+
+    /**
+     * Mapping of scalar types
+     */
+    private Map<String, DateTimeFormatter> dateParsers;
+    private Map<String, DateTimeFormatter> dateExpressions;
+
+    /**
+     * The base template group
+     */
+    private STGroup group;
+
+    /**
+     * Create a new string template engine using the
+     *
+     * @param compilerView
+     */
+    public StringTemplateEngine(RepositoryView compilerView) throws IOException {
+        // Load the root template group that should contain
+        // all the child templates
+        this.group = new STGroupRepository(compilerView, "compiler.stg", '<', '>');
+
+        // Register <escape()> formatter
+        this.group.registerRenderer(String.class, new FormattersRenderer());
+        this.group.defineTemplate("escape", "value", "<value; format=\"_escapeFormatter\">");
+        this.group.defineTemplate("capitalize", "value", "<value; format=\"_capitalizeFormatter\">");
+
+        // Load the language settings
+        Ini ini = new Ini();
+        ini.load(compilerView.openFile("config/language.ini"));
+        String arrayTemplate = ini.get("structures", "array");
+
+        // Load scalar types
+        this.scalarTypes = ini.get("scalar_types");
+
+        // Locate date expression formatters
+        Map<String, String> dateSection = ini.get("date");
+        if (dateSection == null) dateSection = new HashMap<>();
+        this.dateParsers = new HashMap<>();
+        this.dateExpressions = new HashMap<>();
+        this.dateZone = ZoneId.of(dateSection.getOrDefault("timezone", "Europe/Berlin"));
+
+        // Prepare defaults
```

Comment:
I don't quite get why w define date parsers here. When are we parsing dates? In the deserialization?

Reply:
This defines how the date literals reach the compiler. Let me elaborate:

* Consider a type of `type: datetime`
* The engine is going to define a `defaultValue` expression, as a `ValueArguments` instance
    * If the date is an expression that can parsed by snakeyaml, this value is going to be stored internally as a date object
    * If the date is a string that cannot be parsed, the date parser is used instead. The parser depends on the type of the date, so **this** is where the date parsers are used.
* Through the `ValueArguments` interface the user has access to:
    * The raw date instance and it's properties
    * A stringified expression of the date. For **this** value you need the date formatters

Reply:
> If the date is a string that cannot be parsed
If I understand the RAML spec correctly this should not happen and it should be fine to throw an error for unparsable default value.